### PR TITLE
[TwigBundle] Warmup twig templates in non-standard paths

### PR DESCRIPTION
**Pull request by kbond**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12507 |
| License | MIT |
| Doc PR | symfony/symfony-docs#5391 |

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 1]
```diff
@@ -78,6 +78,8 @@ public function load(array $configs, ContainerBuilder $container)
             $container->addResource(new DirectoryResource($path));
         }
 
+        $container->setParameter('twig.paths', $config['paths']);
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: We do not create parameters for such things. Instead, you need to inject the value directly via `replaceArgument` and add a comment in the XML about the value that is going to be injected (have a look at some other examples in the framework).

File: src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php [hunk 2]
```diff
@@ -27,14 +29,16 @@ class TemplateCacheCacheWarmer implements CacheWarmerInterface
...
 
     /**
      * Constructor.
      *
      * @param ContainerInterface      $container The dependency injection container
      * @param TemplateFinderInterface $finder    The template paths cache warmer
+     * @param array                   $paths     Additional twig paths to warm
      */
-    public function __construct(ContainerInterface $container, TemplateFinderInterface $finder)
+    public function __construct(ContainerInterface $container, TemplateFinderInterface $finder, array $paths)
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: You should define a default value of `array()` to keep BC.

**Comment by kbond** [comment]
Those changes were made.

File: src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php [hunk 3]
```diff
@@ -27,14 +29,16 @@ class TemplateCacheCacheWarmer implements CacheWarmerInterface
 {
     protected $container;
     protected $finder;
+    protected $paths;
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: could it be private actually ?

Comment (kbond) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: I typically would make it private but was following the convention of the file. Should I make it private then?

Comment (stof) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: yes please. Any new property should start as being private. We only switch to protected when we decide it should be a supported extension point (and I have no idea why we made the other properties protected btw, but changing it is a BC break)

Comment (kbond) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: I see. This change was made.

File: src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php [hunk 4]
```diff
@@ -75,4 +86,29 @@ public function isOptional()
...
+     */
+    private function findTemplatesInFolder($namespace, $dir)
+    {
+        $templates = array();
+
+        if (is_dir($dir)) {
+            $finder = new Finder();
+            foreach ($finder->files()->followLinks()->in($dir) as $file) {
+                $templates[] = new TemplateReference(
+                    sprintf('@%s/%s', $namespace, $file->getRelativePathname()),
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: you need to handle the case of no namespace too

Comment (kbond) [inline src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php]: You can have a `null` or empty namespace?

So for config that looks like?

```yaml
twig:
    paths:
        "/path/to/templates": `
```

Should it be like this?

```php
$namespace ? sprintf('@%s/%s', $namespace, $file->getRelativePathname()) : $file->getRelativePathname()
```

**Comment by kbond** [comment]
Any chance this can make it into 2.7?

**Comment by jakzal** [comment]
@kbond we don't accept new features in 2.7 anymore. It'll be released in few days.

**Comment by fabpot** [comment]
@kbond Can you rebase on current 2.7 and fix the small CS issue?

**Comment by kbond** [comment]
Done.

**Comment by fabpot** [comment]
:+1:

**Comment by xabbuh** [comment]
:+1:  (though I would rename `findTemplatesInFolder()` to `findTemplatesInDirectory()`)

**Comment by fabpot** [comment]
Thank you @kbond.

**Pull request closed without merging**